baybe/searchspace/_annotated.py

```diff
@@ -0,0 +1,48 @@
+"""Search spaces with metadata."""
+
+import pandas as pd
+from attrs import asdict, define, field
+from typing_extensions import Self, override
+
+from baybe.searchspace import SubspaceDiscrete
+from baybe.utils.boolean import eq_dataframe
+
+
+@define
+class AnnotatedSubspaceDiscrete(SubspaceDiscrete):
```

Comment:
now if we accept that we dont worry about copy operations for now, ist it more straighforward to put this fifltering into the campaign and simply pass an evolved `SearchSpace` via a reduced `SubspaceDiscrete`.
seems like this class is a bit obsolete

Reply:
Sorry, wall of text 📃 But I think these could be some neat ideas and are worth discussing (not necessarily here, though, but with a real coffee/whiteboard):

---

Let me elaborate my plan and then you tell me what you think about it. I've thought quite a few steps ahead and would appreciate to hear your opinion:
1. At the moment, it can't just *yet* be removed, because the `Campaign` alone does not know what is to be removed – it still depends on the `allow_*` flags which are currently one level deeper at the `Recommender`, so the decision must be postponed. That postponing is exactly the purpose of the `Annotated` class. The only way how this *could* be done at the moment is if the `Campaign` would actually read the flags from the *next* recommender, which I don't think is very elegant. It's basically the "Version 3b" described [here](https://github.com/emdgroup/baybe/pull/412#discussion_r1810874618), which I voted against.
2. Instead, I am already working on the follow-up PR implementing "Version 3a", which also moves the `allow_*` flags to the campaign, making it the ruler of metadata (it becomes the only place where this stuff is handled, which to me ultimately seems the nice/correct solution). With that, I'll replace the `AnnotatedSubspaceDiscrete` class with `FilteredSubspaceDiscrete` class in a first step. The difference here is that the latter is also unaware of metadata and just implements a generic interface for lazy candidate removal. One reason is that it'll make the PR a bit smaller but more importantly:
3. We should first discuss in a third PR / baybathon whether we might actually want to keep the class (design will be such that removal is backward-compatible). Because I see quite some potential benefits of having it as an additional layer for lazy operations. That is because I already see (already at the horizon) an upcoming `SearchSpaceProtocol` that will make it really easy for us to finalize the complete switch to `polars` but in a super flexible way that also allows other backends like databases / KDtrees etc. The `Filtered` class might come in very handy here because in those other backends, it might be not as trivial to convert the lazy representation into an explicit one other than through iteration. For the pandas backend, we'd still be able to do something like `searchspace.filter(lazy=True/False)` or `searchspace.filter(lazy=True).collect()` where `filter(lazy=True)` yields a `Filtered` object, whose `collect` method then turns it into an actual filtered space.

Reply:
I didn't read all of this but the first point about the unavailable allow flags makes sense, expectation about further evolvement was provided